utils/self-extr-exec/compressor.cpp

```diff
@@ -0,0 +1,341 @@
+#include <cstring>
+#include <iostream>
+#include <zstd.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdlib.h>
+
+/*
+Overview of compression:
+     ______________________
+    |     Decompressor     |
+    |----------------------|
+    |   Compressed file 1  |
+    |   Compressed file 2  |
+    |         ...          |
+    |----------------------|
+    |   Info about 1 file  |
+    |   Info about 2 file  |
+    |         ...          |
+    |----------------------|
+    |      Metadata        |
+    |______________________|
+*/
+
+/* 
+Metadata contains:
+    1) number of files to support multiple file compression
+    2) start_of_files_data to know start of files metadata
+    3) end of binary to know start of compressed data
+    4) uncompressed data size
+*/
+struct MetaData
+{
+    size_t number_of_files     = 0;
+    size_t start_of_files_data = 0;
+};
+
+/// Information about each file for correct extraction.
+/// Each file data is followed by name of file
+/// with length equals to name_length.
+struct FileData
+{
+    size_t start             = 0;
+    size_t end               = 0;
+    size_t name_length       = 0;
+    size_t uncompressed_size = 0;
+};
+
+/// Main compression part
+int doCompress(char * input, char * output, off_t & in_offset, off_t & out_offset,
+               off_t input_size, off_t output_size, ZSTD_CCtx * cctx)
+{
+    size_t compressed_size = ZSTD_compress2(cctx, output + out_offset, output_size, input + in_offset, input_size);
+    if (ZSTD_isError(compressed_size))
+    {
+        std::cout << "Cannot compress block with ZSTD: " + std::string(ZSTD_getErrorName(compressed_size)) << std::endl;
+        return 1;
+    }
+    in_offset += input_size;
+    out_offset += compressed_size;
+    return 0;
+}
+
+/// compress data from opened file into output file
+int compress(int in_fd, int out_fd, int level, off_t& pointer, const struct stat& info_in)
+{
+    /// As experiments showed, size of compressed file is 4 times less than clickhouse executable
+    /// Get a little bit more memory to prevent errors with size. 
+    /// For compression this difference will not be huge
+    if (0 != ftruncate(out_fd, pointer + info_in.st_size / 3))
```

Comment:
That's a little bit strange. Looks like if compressed size > 1/3, than we would get segfault (or error)?
Maybe it would be easier just to stream into output file?

Also, I think you can actually map zero pages without allocating memory. I don't exactly know how to properly do it, maybe [this](https://stackoverflow.com/a/51392323) or [this](https://stackoverflow.com/a/44867334) will work. (some combination of mmap, munmup and msync)